publish_process.js

```diff
@@ -0,0 +1,315 @@
+  return raw.trim();
+}
+
+function close() {
+  rl.close();
+  return false;
+}
+
+async function Main() {
+  try {
+    // verify that we are on the master branch.
+    await ensureMaster();
+    await isTagMissing();
+    let {updateChangeLogWith, currentChangeLogVersion} = await changeLogAndVersion();
+
+    const lastPreparedTag = await getLastPreparedTag();
+    if(lastPreparedTag) {
+      // Here we only need to publish the last prepared tag
+      await HandleFinalPublishingProcess(currentChangeLogVersion, lastPreparedTag);
+      return;
+    }
+
+    let newVersion = await nextVersion(currentChangeLogVersion);
+    let mostRecentDafnyRelease = await getMostRecentDafnyRelease().substring(1);
+    let packageObj = await readPackageJson();
+    
+    console.log(`Going to proceed to publish ${newVersion}`);
+    // Get all the commit messages since the last published tag
+    await UpdateChangeLog(currentChangeLogVersion, packageObj, updateChangeLogWith, newVersion);
+    // All clear, we can modify constants.ts and package.json.
+
+    var updatedDafny = await updatePackageJson(packageObj, newVersion, mostRecentDafnyRelease);
+    // Execute npm install to ensure the package lock is up to date
+    console.log("Executing `npm install`...");
+    await execAsync("npm install");
+
+    // Create the new branch and git add all the files modified above
+    console.log("Creating new branch...");
+    const newBranch = `release-${newVersion}`;
+    await execAsync(`git checkout -b ${newBranch}`);
+    await execAsync(`git add ${changeLogFile} ${packageFile} ${constantsFile} ${packageLockFile}`);
+    if(ok(await question(`I made all the necessary edits. Push the changes to the remote repository? ${ACCEPT_HINT}`))) {
+      await execAsync(`git commit -m "Release v${newVersion}${ updatedDafny ? ` (updated Dafny to ${mostRecentDafnyRelease})` : "" }"`);
+      await execAsync(`git push origin ${newBranch}`);
+      console.log("Now, create the pull request by clicking the link below:");
+      console.log(`https://github.com/dafny-lang/ide-vscode/compare/${newBranch}?expand=1`);
+      console.log("When this PR is approved and merged, launch this script again to finish publishing the release.");
+    } else {
+      console.log("Aborting publishing.");
+      return close();
+    }
+
+  } catch(e) {
+    if(e != ABORTED) {
+      throw e;
+    }
+  } finally {
+    close();
+  }
+}
+Main();
+
+
+async function updatePackageJson(packageObj, newVersion, mostRecentDafnyRelease) {
+  packageObj["version"] = newVersion;
+  var previousDafnyVersion = packageObj["contributes"]["configuration"]["properties"]["dafny.preferredVersion"]["enum"][1];
+  var updatedDafny = false;
+  if (previousDafnyVersion != mostRecentDafnyRelease) {
+    if (ok(await question(`The Dafny version in the package.json file (${previousDafnyVersion}) is not the latest (${mostRecentDafnyRelease}). Do you want to update it? ${ACCEPT_HINT}`))) {
```

Comment:
Love that you're automating this part too! 

Food for thought in the future: if we expressed the dependency on the latest known Dafny version in a dummy `.csproj` file, then GitHub's Dependabot automation might be able to manage this for us, since Dafny is now published to nuget as a dotnet tool: https://medium.com/@nickfane/integrating-dependabot-into-your-net-core-project-on-github-3e3024bd3394

Mind you it's a little more complicated in this case since we have to maintain a LIST of known versions, but worth thinking about.